crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -499,6 +524,103 @@ impl<'a> TimelineEventHandler<'a> {
+                    original_sender = ?event_item.sender(), edit_sender = ?self.ctx.sender,
+                    "Edit event applies to another user's timeline item, discarding"
+                );
+                return None;
+            }
+
+            let poll_state = if let TimelineItemContent::Poll(poll_state) = &event_item.content() {
+                poll_state
+            } else {
+                info!(
+                    original_sender = ?event_item.sender(), edit_sender = ?self.ctx.sender,
+                    "Can't edit a poll that is not of type TimelineItemContent::Poll, discarding"
+                );
+                return None;
+            };
+
+            let new_content =
+                if let Ok(edited_poll_state) = poll_state.edit(&replacement.new_content) {
+                    TimelineItemContent::Poll(edited_poll_state)
+                } else {
+                    info!(
+                        original_sender = ?event_item.sender(), edit_sender = ?self.ctx.sender,
+                        "This poll can't be edited, discarding"
+                    );
+                    return None;
+                };
+
+            let edit_json = match &self.ctx.flow {
+                Flow::Local { .. } => None,
+                Flow::Remote { raw_event, .. } => Some(raw_event.clone()),
+            };
+
+            trace!("Applying edit");
+            Some(event_item.with_content(new_content, edit_json))
+        });
+    }
+
+    fn handle_poll_start(&mut self, c: UnstablePollStartEventContent, should_add: bool) {
+        let mut poll_state = PollState::new(c);
+        // TODO: If the flow is not remote we're not able to apply the pending events. Is this a problem?
+        if let Flow::Remote { event_id, .. } = self.ctx.flow.clone() {
+            self.state.poll_pending_events.apply(&event_id, &mut poll_state);
+        };
+        self.add(should_add, TimelineItemContent::Poll(poll_state));
+    }
+
+    fn handle_poll_response(&mut self, c: UnstablePollResponseEventContent) {
+        update_timeline_item!(
+            self,
+            &c.relates_to.event_id,
+            found: |event_item| match event_item.content() {
+                TimelineItemContent::Poll(poll_state) => Some(event_item.with_content(
+                    TimelineItemContent::Poll(poll_state.add_response(
+                        &self.ctx.sender,
+                        &self.ctx.timestamp,
+                        &c,
+                    )),
+                    None,
+                )),
+                _ => None,
+            },
+            not_found: || {
+                self.state.poll_pending_events.add_response(
+                    &c.relates_to.event_id,
+                    &self.ctx.sender,
+                    &self.ctx.timestamp,
+                    &c,
+                );
+            }
+        );
+    }
+
+    fn handle_poll_end(&mut self, c: UnstablePollEndEventContent) {
+        update_timeline_item!(
```

Comment:
We should check first if the user has the right to end the poll. According to the MSC it is if its the creator of the poll, or they have the right to redact any event.

Reply:
> We should check first if the user has the right to end the poll. According to the MSC it is if its the creator of the poll, or they have the right to redact any event.

@jplatte on this regard, can you point me to some existing code that deals with checking power levels so I can "borrow" (rust pun intended) some wisdom from it?

Reply:
https://github.com/matrix-org/matrix-rust-sdk/blob/4493cbf0ace49acf1667dbce6efb24ff8312551c/crates/matrix-sdk/src/room/mod.rs#L2025